src/main.rs

```diff
@@ -192,107 +227,150 @@ impl<'a> Interpreter<'a> {
+                        max_local = max(max_local, stack_prefix + idx);
                     }
 
+                    call_tree = call_tree.push(FrameInfo {
+                        func: new_func.clone(),
+                        back_ref: frame.back_ref.clone(),
+                        offset: stack_prefix,
+                    });
+
                     self.frames.push(frame);
 
                     func = new_func;
                     pc = 0;
                     continue;
-                },
+                }
 
                 Return => {
-                    let frame = self.frames.pop();
+                    stack_prefix = stack_prefixes.pop().unwrap();
 
+                    let frame = self.frames.pop();
                     if self.frames.is_empty() {
                         break;
                     }
 
+                    call_tree = call_tree.pop().unwrap();
+
                     let (f, rpc) = frame.unwrap().back_ref;
                     func = f;
                     pc = rpc;
                     continue;
-                },
+                }
 
                 Cmp(how) => self.do_cmp(how),
 
                 Jump(target) => {
                     pc = target;
-                    // don't trace
+                    // skip trace
                     continue;
                 }
 
                 JumpIfFalse(target) => {
-                    let b: bool = self.pop();
+                    let b: bool = self.stack.pop_into();
                     if !bool::from(b) {
                         pc = target;
                     }
 
-                    trace.push(Guard(b));
+                    let guard = Guard {
+                        condition: b,
+                        frame: call_tree.clone(),
+                        // reverse `pc+=1` above
+                        pc: pc - 1,
+                    };
+                    trace.push(TraceInstruction::Guard(guard));
                     continue;
-
                 }
 
                 _ => panic!("TODO: {:?}", instr),
             }
 
-            trace.push(instr.clone());
+            trace.push(TraceInstruction::from(instr));
         }
 
-        println!("{:?}", trace);
+        info!(target: "trace", "{:?}", trace);
 
-        (func, pc)
+        (func,
+         pc,
+         Trace {
+            trace: trace,
+            locals: max_local + 1,
+        })
     }
 
     fn run(&mut self) {
         use Instruction::*;
 
+        // `main` function has to exist
         let main = self.get_fn("main");
 
-        let mut pc = 0;
-
-        self.frames.push(CallFrame::for_fn(&main, (&main, 0)));
+        // a bit awkward, main would return to main
+        // maybe it would be better to have Option as back_ref
+        self.frames.push(CallFrame::for_fn(&*main, (main.clone(), 0)));
 
         let mut func = main;
+        let mut pc = 0;
+        let mut traces: BTreeMap<usize, Trace> = BTreeMap::new();
```

Comment:
AFAIK the default hashing algorithm used in rust is (or was) quite slow, but secure. But to be honest, I don't really know why I choose btreemap here. What would be the advantages of using a hashmap?